src/parquet/util/input.cc

```diff
@@ -54,23 +58,47 @@ void LocalFileSource::Close() {
   CloseFile();
 }
 
-void LocalFileSource::CloseFile() {
+void LocalFileSource::CloseFile() const {
   if (is_open_) {
     fclose(file_);
     is_open_ = false;
   }
 }
 
-void LocalFileSource::Seek(int64_t pos) {
-  fseek(file_, pos, SEEK_SET);
-}
-
 int64_t LocalFileSource::Size() const {
+  if (size_ < 0) {
+    int64_t current_position = Tell();
+    SeekFile(0, SEEK_END);
+    size_ = Tell();
+    SeekFile(current_position);
+  }
   return size_;
 }
 
+void LocalFileSource::SeekFile(int64_t pos, int origin) const {
+  if (origin == SEEK_SET && pos >= size_) {
+    std::stringstream ss;
+    ss << "Cannot seek to position " << pos << " in file [" << path_
+       << "] of size " << size_;
+    throw ParquetException(ss.str());
+  }
```

Comment:
Some implementations of fseek() will not produce an error if you try to seek past the end of the file. Hence, we should track this explicitly.